resteasy-reactive: ParamConverter+Kotlin+ResteasyReactice+List<SomeClass> leads to java.lang.ClassNotFoundException on startup (Regression since 2.12)

### Describe the bug

__Regression Quarkus 2.11 -> Quarkus 2.12__

The following example leads to an error on startup (reproducer attached):

```kotlin
@Provider
class XyzConverterProvider : ParamConverterProvider {

    override fun <T> getConverter(
        rawType: Class<T>,
        genericType: Type,
        annotations: Array<Annotation>
    ): ParamConverter<T>? {
        return null
    }
}

enum class MyEnum {
    A, B
}

@Path("/hello")
class GreetingResource {

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    fun hello(
        @QueryParam("value") values: List<MyEnum>
    ) = values.toString()
}
```

```bash
2022-09-13 10:01:31,055 ERROR [io.qua.run.boo.StartupActionImpl] (Quarkus Main Thread) Error running Quarkus: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:103)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.ExceptionInInitializerError
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.lang.Class.newInstance(Class.java:584)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:66)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:41)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:120)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
	... 6 more
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
	... 15 more
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: ? extends org.acme.MyEnum
	at org.jboss.resteasy.reactive.common.util.DeploymentUtils.loadClass(DeploymentUtils.java:43)
	at org.jboss.resteasy.reactive.server.core.startup.RuntimeResourceDeployment.smartInitParameterConverter(RuntimeResourceDeployment.java:506)
	at org.jboss.resteasy.reactive.server.core.startup.RuntimeResourceDeployment.buildResourceMethod(RuntimeResourceDeployment.java:349)
	at org.jboss.resteasy.reactive.server.core.startup.RuntimeDeploymentManager.deploy(RuntimeDeploymentManager.java:116)
	at io.quarkus.resteasy.reactive.server.runtime.ResteasyReactiveRecorder.createDeployment(ResteasyReactiveRecorder.java:174)
	at io.quarkus.deployment.steps.ResteasyReactiveProcessor$setupDeployment1557341086.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.ResteasyReactiveProcessor$setupDeployment1557341086.deploy(Unknown Source)
	... 16 more
Caused by: java.lang.ClassNotFoundException: ? extends org.acme.MyEnum
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:490)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:440)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:490)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:440)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at org.jboss.resteasy.reactive.common.util.DeploymentUtils.loadClass(DeploymentUtils.java:41)
	... 22 more
```

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

[reproducer.zip](https://github.com/quarkusio/quarkus/files/9555095/reproducer.zip)


### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

Java 11

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

Quarkus 2.12.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven

### Additional information

The roblem was introduced with #27529.

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/providers/ParamConverterTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/providers/ParamConverterTest.java
@@ -73,11 +73,11 @@ public class ParamConverterTest {
 
         @Path("list")
         @GET
-        public String list(@QueryParam("id") List<UUID> uuids) {
+        public String list(@QueryParam("id") List<? extends UUID> uuids) {
             return join(uuids.stream());
         }
 
-        private static String join(Stream<UUID> uuids) {
+        private static String join(Stream<? extends UUID> uuids) {
             return uuids.map(UUID::toString).collect(Collectors.joining(","));
         }
     }

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/startup/RuntimeResourceDeployment.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/startup/RuntimeResourceDeployment.java
@@ -8,6 +8,7 @@ import java.lang.annotation.Annotation;
 import java.lang.reflect.Method;
 import java.lang.reflect.ParameterizedType;
 import java.lang.reflect.Type;
+import java.lang.reflect.WildcardType;
 import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -503,7 +504,27 @@ public class RuntimeResourceDeployment {
             if (genericParameterTypes[i] instanceof ParameterizedType) {
                 Type[] genericArguments = ((ParameterizedType) genericParameterTypes[i]).getActualTypeArguments();
                 if (genericArguments.length == 1) {
-                    quarkusConverter.init(paramConverterProviders, loadClass(genericArguments[0].getTypeName()),
+                    String genericTypeClassName = null;
+                    Type genericType = genericArguments[0];
+                    if (genericType instanceof Class) {
+                        genericTypeClassName = ((Class<?>) genericType).getName();
+                    } else if (genericType instanceof WildcardType) {
+                        WildcardType genericTypeWildcardType = (WildcardType) genericType;
+                        Type[] upperBounds = genericTypeWildcardType.getUpperBounds();
+                        Type[] lowerBounds = genericTypeWildcardType.getLowerBounds();
+                        if ((lowerBounds.length == 0) && (upperBounds.length == 1)) {
+                            Type genericTypeUpperBoundType = upperBounds[0];
+                            if (genericTypeUpperBoundType instanceof Class) {
+                                genericTypeClassName = ((Class<?>) genericTypeUpperBoundType).getName();
+                            }
+                        }
+                    }
+                    //TODO: are there any other cases we can support?
+                    if (genericTypeClassName == null) {
+                        throw new IllegalArgumentException(
+                                "Unable to support parameter converter with type: '" + genericType.getTypeName() + "'");
+                    }
+                    quarkusConverter.init(paramConverterProviders, loadClass(genericTypeClassName),
                             genericArguments[0],
                             parameterAnnotations[i]);
                     return;

